File: features/task/src/main/java/com/nivasafinance/features/task/entity/Task.java
package com.nivasafinance.features.task.entity;

import com.nivasafinance.common.audit.AuditableEntity;
import com.nivasafinance.common.enums.EntityType;
import io.hypersistence.utils.hibernate.type.json.JsonType;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.Table;
import lombok.AccessLevel;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.EqualsAndHashCode;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.hibernate.annotations.JdbcTypeCode;
import org.hibernate.annotations.Type;
import org.hibernate.type.SqlTypes;

import java.time.LocalDateTime;
import java.util.UUID;

@Entity
@Table(name = "n_tasks")
@NoArgsConstructor
@AllArgsConstructor
@Data
@EqualsAndHashCode(callSuper = true)
public class Task extends AuditableEntity {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @JdbcTypeCode(SqlTypes.UUID)
    @Column(name = "task_identifier", nullable = false, unique = true, updatable = false, columnDefinition = "UUID")
    @Setter(AccessLevel.NONE)
    private UUID taskIdentifier;

    @Column(name = "task_config_key", nullable = false, length = 100)
    private String taskConfigKey;

    @Column(name = "name", length = 100)
    private String name;

    @Column(name = "assigned_to", length = 255)
    private String assignedTo;

    @Column(name = "due_at")
    private LocalDateTime dueAt;

    @Column(name = "outcome", length = 100)
    private String outcome;

    @Type(JsonType.class)
    @JdbcTypeCode(SqlTypes.JSON)
    @Column(name = "outcome_details", columnDefinition = "jsonb")
    private OutcomeDetails outcomeDetails;

    @Type(JsonType.class)
    @JdbcTypeCode(SqlTypes.JSON)
    @Column(name = "task_details", columnDefinition = "jsonb")
    private TaskDetails taskDetails;

    @jakarta.persistence.PrePersist
    void prePersist() {
        if (!com.nivasafinance.common.utils.ValidationUtils.isNonNull(taskIdentifier)) {
            taskIdentifier = UUID.randomUUID();
        }
    }

    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    @Builder
    public static class OutcomeDetails {
        private String remarks;
        private LocalDateTime completedAt;
        private String completedBy;
        private String rescheduleReasonCodeValueKey;
    }

    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    @Builder
    public static class TaskDetails {
        private UUID entityId;
        private EntityType entityType;
        private PreferredCallWindow preferredCallWindow;
        private String creatorRemarks;
        private Integer iterationCount;
        private UUID rescheduledFromTaskIdentifier;
        private String rescheduleReasonCodeValueKey;
        private String rescheduledFromTaskRemarks;

        @Data
        @NoArgsConstructor
        @AllArgsConstructor
        @Builder
        public static class PreferredCallWindow {
            private LocalDateTime start;
            private LocalDateTime end;
        }
    }
}

File: features/tasks/src/main/java/com/nivasafinance/features/tasks/entity/Task.kt
package com.nivasafinance.features.tasks.entity

import com.nivasafinance.common.audit.AuditableEntity
import com.nivasafinance.features.tasks.enum.TaskStatus
import jakarta.persistence.Column
import jakarta.persistence.Entity
import jakarta.persistence.EnumType
import jakarta.persistence.Enumerated
import jakarta.persistence.GeneratedValue
import jakarta.persistence.GenerationType
import jakarta.persistence.Id
import jakarta.persistence.Table
import java.time.LocalDateTime
import java.util.UUID

@Entity
@Table(name = "tasks")
class Task : AuditableEntity() {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    val id: UUID? = null

    @Column(name = "task_definition_key", nullable = false)
    var taskDefinitionKey: String = ""

    @Column(name = "name", nullable = false)
    var name: String = ""

    @Column(name = "description", nullable = true)
    var description: String? = null

    @Enumerated(EnumType.STRING)
    @Column(name = "status", nullable = false)
    var status: TaskStatus = TaskStatus.TODO

    @Column(name = "outcome", nullable = true)
    var outcome: String? = null

    @Column(name = "assigned_to", nullable = true)
    var assignedTo: String? = null

    @Column(name = "due_at", nullable = true)
    var dueAt: LocalDateTime? = null

    @Column(name = "completed_at", nullable = true)
    var completedAt: LocalDateTime? = null

    @Column(name = "completed_by", nullable = true)
    var completedBy: String? = null
}
